### PR TITLE
HTTPCLIENT-2277: Redesign of the internal cache serialization format

### PR DESCRIPTION
**Pull request by ok2c**
@arturobernalg This change-set builds upon your previous work on `HttpByteArrayCacheEntrySerializer`. What has changed:

1. Cache entries have a version now. In the future we can change the underlying serialization format without breaking binary compatibility with the previous versions of storage backends
2. Cache entries now store request details (method, request URI and headers) as well.
3. There is now cache entry specific metadata that is not a part of request or response messages (cache key, cache size, etc)
4. No more escaped headers. 
5. Lots of test code cleanups

Please review and let me know if you find anything disagreeable.

### CODE REVIEW

**Comment by arturobernalg** [comment]
@ok2c 
I'll certainly examine this. Kindly grant me a couple of days to review all the changes in the code.

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheUpdateHandler.java [hunk 1]
```diff
@@ -123,18 +128,29 @@ public HttpCacheEntry updateParentCacheEntry(
...
-            return entry.getHeaders();
+    private static HeaderGroup headers(final Iterator<Header> it) {
+        final HeaderGroup headerGroup = new HeaderGroup();
+        while (it.hasNext()) {
+            headerGroup.addHeader(it.next());
         }
+        return headerGroup;
+    }
+
+    private HeaderGroup mergeHeaders(final HttpCacheEntry entry, final HttpResponse response) {
```
Comment (arturobernalg) [inline httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheUpdateHandler.java:151]: Currently, the method iterates over the response headers twice: once to remove matching headers from the cache, and once to add the response headers. In both iterations, the Content-Encoding and Content-Length headers are being skipped. This seems inefficient.

**Review by arturobernalg — CHANGES_REQUESTED** [review]

File: httpclient5-cache/src/test/java/org/apache/hc/client5/http/impl/cache/HttpCacheEntryMatcher.java [hunk 2]
```diff
@@ -92,6 +91,41 @@ public boolean matches(final Object item) {
...
+                return false;
+            }
+        }
+        if (other.hasNext()) {
+            return false;
+        }
+        return true;
+    }
+
+    private static boolean mapEqual(final Map<?, ?> expected, final Map<?, ?> actual) {
```
[the remark this answers is not in the document]
Comment (ok2c) [reply httpclient5-cache/src/test/java/org/apache/hc/client5/http/impl/cache/HttpCacheEntryMatcher.java:117]: > You can first check if the sizes of the maps are equal. 
@arturobernalg Makes sense.
> Then I would use allMatch function
I am not sure I know what you mean. Is this a matcher?

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/HttpByteArrayCacheEntrySerializer.java [hunk 3]
```diff
@@ -230,316 +233,174 @@ public HttpCacheStorageEntry deserialize(final byte[] serializedObject) throws R
...
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("Deserializing cache entry with headers: {}", response.getHeaders());
+            final CharArrayBuffer line = new CharArrayBuffer(DEFAULT_BUFFER_SIZE);
+            checkReadResult(inputBuffer.readLine(line, in));
+            final String versionLine = line.toString();
+            if (!versionLine.equals(HC_CACHE_VERSION_LINE)) {
+                throw new ResourceIOException("Unexpected cache entry version line");
+            }
+            final List<Header> cacheHeaders = new ArrayList<>();
+            while (true) {
```
[the remark this answers is not in the document]
Comment (ok2c) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/HttpByteArrayCacheEntrySerializer.java]: @arturobernalg Corrected.

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/HttpByteArrayCacheEntrySerializer.java [hunk 4]
```diff
@@ -163,57 +125,98 @@ public HttpByteArrayCacheEntrySerializer() {
...
+        final Resource resource = cacheEntry.getResource();
 
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream((resource != null ? (int) resource.length() : 0) + DEFAULT_BUFFER_SIZE)) {
             final SessionOutputBuffer outputBuffer = new SessionOutputBufferImpl(bufferSize);
+            final CharArrayBuffer line = new CharArrayBuffer(DEFAULT_BUFFER_SIZE);
+
+            line.append(HC_CACHE_VERSION_LINE);
+            outputBuffer.writeLine(line, out);
+
+            line.clear();
```
[the remark this answers is not in the document]
Comment (ok2c) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/HttpByteArrayCacheEntrySerializer.java:143]: @arturobernalg I disagree. This is really a matter of taste. I personally find a utility method with 4 arguments much uglier than 5 repetitive trivial lines.

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheUpdateHandler.java (hunk 1 above)
Comment (ok2c) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheUpdateHandler.java:151]: @arturobernalg Bear with me. `Content-Length ` stuff will be gone soon.

**Review by arturobernalg — APPROVED** [review]
The changes look good. I've added a comment on the `mapEqual` method.

File: httpclient5-cache/src/test/java/org/apache/hc/client5/http/impl/cache/HttpCacheEntryMatcher.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ok2c) [reply httpclient5-cache/src/test/java/org/apache/hc/client5/http/impl/cache/HttpCacheEntryMatcher.java:117]: @arturobernalg Ah. Nice! Thank you!

**Pull request merged**